Feature/ab2d 2146 coverage mapping service

wnyffenegger opened the pull request:
**JIRA Tickets:**

[AB2D-2146](https://jira.cms.gov/browse/AB2D-2146) - Coverage Mapping CRUD Service

***Related Tickets***

[AB2D-2110](https://jira.cms.gov/browse/AB2D-2110) - DB Schema Updates
 
### What Does This PR Do?

Create the initial CoverageService for interacting with the tables represented by the CoveragePeriod, CoverageSearchEvent, and Coverage entities. Manages access to these entities and the status progression of a search.

This PR does not address bulk insert or read operations but does address bulk delete operations.

### What Should Reviewers Watch For?

JPA and native queries in the CoverageRepository, CoverageSearchEventRepository, and CoveragePeriodRepository do not have any major flaws that may cause issues.

### Usage/Deployment Instructions

### Impacted External Components

### Database Changes

None. Except service can now enter data into the database.

### Limitations

### Security Implications
<!-- If any boxes are checked, a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence -->

- [ ] This PR adds new software dependencies
- [ ] This PR modifies or invalidates our security controls
- [ ] This PR stores or transmits data that was not stored or transmitted before

<!-- If any boxes are checked, please add @StewGoin as a reviewer, and this PR should not be merged unless/until he also approves it. -->

- [ ] This PR requires additional review of its security implications for other reasons

### What Needs to Be Merged and Deployed Before this PR?

### Submitter Checklist

- [ ] This PR includes any required documentation changes, (`README`, changelog / release notes, website).
- [ ] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments.
- [ ] Code checked for PHI/PII exposure

nflearl commented on common/src/test/java/gov/cms/ab2d/common/EntityUtils.java:
```diff
@@ -0,0 +1,72 @@
+import java.time.OffsetDateTime;
+import java.util.Random;
+
+public class EntityUtils {
```
I don't like the concept of this class.  Put a static create method on each domain class and then where used, it's simply a repo.save(Domain.create(...)) call.

adaykin replied:
These should be in the service classes for each entity it's related to, not the domain class. They use the repository, which is a bean, the repo is already in the service class.

adaykin replied:
Actually I misread this, it's for testing purposes. I think it's fine to have it stay here, although I think there's already a bean that does some data creation

adaykin replied:
DataSetup.java is the bean if you can use it. That way you don't have to pass the repo into every method.

wnyffenegger replied:
I'll move everything to DataSetup instead.

nflearl commented on common/src/test/java/gov/cms/ab2d/common/service/CacheServiceImplTest.java:
```diff
@@ -57,22 +55,22 @@
     void setUp() {
         coverageRepo.deleteAll();
         coverageSearchEventRepo.deleteAll();
-        coverageSearchRepo.deleteAll();
+        coveragePeriodRepo.deleteAll();
```
Move to teardown()

nflearl commented on common/src/test/java/gov/cms/ab2d/common/service/CoverageServiceImplTest.java:
```diff
@@ -0,0 +1,433 @@
+
+    @AfterEach
+    public void cleanUp() {
+        coverageRepo.deleteAll();
+        coverageSearchEventRepo.deleteAll();
+        coveragePeriodRepo.deleteAll();
+        contractRepo.deleteAll();
```
Move to teardown()

nflearl reviewed (changes requested):
See inline comments.

adaykin commented on common/src/main/java/gov/cms/ab2d/common/repository/CoverageRepository.java:
```diff
@@ -11,10 +12,23 @@
 
+    int countByCoverageSearchEvent(CoverageSearchEvent searchEvent);
+
+    @Query(value = "SELECT COUNT(*) FROM (" +
```
That might be difficult to do with this particular query, the example listed is a lot simpler than this query.

adaykin commented on common/src/main/java/gov/cms/ab2d/common/repository/CoverageSearchEventRepository.java:
```diff
@@ -1,7 +1,25 @@
 
 public interface CoverageSearchEventRepository extends JpaRepository<CoverageSearchEvent, Long> {
+
+    @Query(value = "SELECT * FROM event_bene_coverage_search_status_change as ebc " +
+            "   WHERE ebc.bene_coverage_period_id = ? AND ebc.new_status = 'IN_PROGRESS'" +
+            "   ORDER BY ebc.created DESC LIMIT 1 OFFSET 1",
+            nativeQuery = true)
+    Optional<CoverageSearchEvent> findPreviousSearch(int searchId);
+
+    @Query(value = "SELECT * FROM event_bene_coverage_search_status_change as ebc " +
+            "   WHERE ebc.bene_coverage_period_id = ? AND ebc.new_status = 'IN_PROGRESS'" +
+            "   ORDER BY ebc.created DESC LIMIT 1",
+            nativeQuery = true)
+    Optional<CoverageSearchEvent> findCurrentSearch(int searchId);
```
I think the SQL is actually more readable than the criteria builder. If you can do named parameters instead of ? it may be a better way of being "type safe".

adaykin commented on common/src/main/java/gov/cms/ab2d/common/service/CoverageServiceImpl.java:
```diff
@@ -0,0 +1,258 @@
+
+    @Transactional
+    @Override
+    public CoverageSearchEvent insertCoverage(int searchId, long searchEventId, Collection<String> beneficiaryIds) {
```
Not sure you need transactional at the method level in all these places if it's at the class level.

adaykin commented on common/src/main/java/gov/cms/ab2d/common/service/CoverageServiceImpl.java:
```diff
@@ -0,0 +1,258 @@
+        CoveragePeriod period = coveragePeriodRepo.getOne(searchId);
+
+        if (period.getStatus() != JobStatus.IN_PROGRESS) {
+            throw new InvalidJobStateTransition("");
```
Let's provide some sort of message here.

adaykin commented on common/src/main/java/gov/cms/ab2d/common/service/CoverageServiceImpl.java:
```diff
@@ -0,0 +1,258 @@
+
+    @Transactional
+    @Override
+    public CoverageSearchEvent insertCoverage(int searchId, long searchEventId, Collection<String> beneficiaryIds) {
```
Same thing here with Transactional at the class and method level.

wnyffenegger replied:
Yeah it shouldn't be at the class level, it should be at the method level.

adaykin commented on common/src/main/java/gov/cms/ab2d/common/service/CoverageServiceImpl.java:
```diff
@@ -0,0 +1,258 @@
+
+    @Transactional
+    @Override
+    public void deletePreviousSearch(int searchId) {
```
Where does this get called from?

wnyffenegger replied:
@adaykin this is actually up for debate. Basically we want all the previous search results in the database until the newest search has completed. The reason we want this is to diff our results to log any changes. Laila wants to log those changes so we can get a handle on when.

The thread actually doing the search could do these operations or deletePreviousSearch could be called in `CoverageServiceImpl.completeCoverageSearch()`

adaykin replied:
Ok so it's planned to be added in the future? Just wanted to make sure you hadn't forgotten to call it somewhere.

adaykin commented on common/src/main/java/gov/cms/ab2d/common/service/CoverageServiceImpl.java:
```diff
@@ -0,0 +1,258 @@
+    }
+
+    @Override
+    public CoverageSearchDiff searchDiff(int searchId) {
```
Same here about where it's getting called.

wnyffenegger replied:
Not being called yet but eventually use to keep statistics about what was changed between searches.

wnyffenegger commented:
@adaykin @nflearl 

* Replaced all lazy loading of `CoveragePeriod` and `CoverageSearchEvent` with eager loading.
* Discovered bug related to deletion of a contract due to foreign key relationship with sponsor. Affects all contractRepo.deleteAll statemens
* Added manual cascade to force a contract to be removed from a sponsor's list of contracts when the contract is deleted as fix for that bug.
* Discovered many other test issues caused by correctly deleting contracts and repos that needed to be cleaned up

wnyffenegger commented on common/src/main/java/gov/cms/ab2d/common/model/CoverageSearchEvent.java:
```diff
@@ -17,7 +20,8 @@
 
-    @ManyToOne
+    // Do not run a
+    @ManyToOne(fetch = FetchType.EAGER)
```
Removed eager loading actually. Artifact from transactional issues.

nflearl reviewed (changes requested):
Mainly want the todo comment added, that was quite a discussion that bears a brief capture.
